File: frontend/src/runtime/three/ParticleSimulation.ts
import * as THREE from 'three';
import type { Emitter, Particle } from '../../types';
import { sampleColor, sampleCurve } from '../../utils/curveUtils';
import { clamp } from '../../utils/math';
import { SeededRandom } from './SeededRandom';
import { vec3Pool } from '../../utils/mathPool';

export interface RuntimeEmitter {
  emitter: Emitter;
  particles: Particle[];
  freeSlots: number[];
  aliveCount: number;
  spawnAcc: number;
  localTime: number;
  rng: SeededRandom;
  dirty: boolean;
}

export function createParticles(count: number): Particle[] {
  return Array.from({ length: count }, () => ({
    alive: false, age: 0, life: 0,
    px: 0, py: 0, pz: 0,
    vx: 0, vy: 0, vz: 0,
    rot: 0, spin: 0, baseSize: 1,
    col: { r: 1, g: 1, b: 1, a: 1 },
    frame: 0,
    boneOx: 0, boneOy: 0, boneOz: 0,
    stretchRot: 0, stretch: 0,
  }));
}

export function resetRuntime(runtime: RuntimeEmitter): void {
  const p = runtime.particles;
  for (let i = 0; i < p.length; i++) p[i].alive = false;
  runtime.freeSlots = Array.from({ length: runtime.particles.length }, (_, i) => runtime.particles.length - 1 - i);
  runtime.aliveCount = 0;
  runtime.spawnAcc = 0;
  runtime.localTime = 0;
  runtime.rng = new SeededRandom(runtime.emitter.uid * 2654435761);
  runtime.dirty = false;
}

export function spawnParticle(runtime: RuntimeEmitter): void {
  const emitter = runtime.emitter;
  const idx = runtime.freeSlots.pop();
  if (idx === undefined) return;
  const slot = runtime.particles[idx];
  runtime.aliveCount++;

  slot.alive = true;
  slot.age = 0;
  slot.life = Math.max(0.01, emitter.life + runtime.rng.centered(emitter.lifeRnd));

  const radius = Math.max(0.01, emitter.shapeRadius || 0.35);
  let ox = 0, oy = 0, oz = 0;

  if (emitter.shape === 'box') {
    ox = runtime.rng.centered(radius * 0.7);
    oy = runtime.rng.centered(radius * 0.7);
    oz = runtime.rng.centered(radius * 0.7);
  } else if (emitter.shape === 'sphere' || emitter.shape === 'spherevol') {
    const r = emitter.shape === 'spherevol' ? Math.cbrt(runtime.rng.next()) * radius : radius;
    const theta = runtime.rng.next() * Math.PI;
    const phi = runtime.rng.next() * Math.PI * 2;
    ox = r * Math.sin(theta) * Math.cos(phi);
    oy = r * Math.cos(theta);
    oz = r * Math.sin(theta) * Math.sin(phi);
  } else if (emitter.shape === 'ring' || emitter.shape === 'disc' || emitter.shape === 'cone') {
    const angle = runtime.rng.next() * Math.PI * 2;
    const r = emitter.shape === 'ring' ? radius
      : emitter.shape === 'cone' ? Math.sqrt(runtime.rng.next()) * radius * 0.7
      : Math.sqrt(runtime.rng.next()) * radius;
    ox = r * Math.cos(angle);
    oz = r * Math.sin(angle);
  }

  slot.boneOx = 0;
  slot.boneOy = 0;
  slot.boneOz = 0;
  slot.px = ox;
  slot.py = oy;
  slot.pz = oz;

  const yaw = THREE.MathUtils.degToRad(emitter.dirYaw);
  const pitch = THREE.MathUtils.degToRad(emitter.dirPitch);
  const direction = vec3Pool.acquire();
  direction.set(Math.cos(pitch) * Math.sin(yaw), Math.sin(pitch), Math.cos(pitch) * Math.cos(yaw)).normalize();
  const axis = vec3Pool.acquire();
  axis.set(runtime.rng.centered(), runtime.rng.centered(), runtime.rng.centered()).normalize();
  direction.applyAxisAngle(axis, runtime.rng.centered(THREE.MathUtils.degToRad(emitter.spread) * 0.5));
  const speed = Math.max(0, emitter.speed + runtime.rng.centered(emitter.speedRnd));
  slot.vx = direction.x * speed;
  slot.vy = direction.y * speed;
  slot.vz = direction.z * speed;
  vec3Pool.release(axis);
  vec3Pool.release(direction);

  if (emitter.rotType === 'NONE') {
    slot.rot = 0;
    slot.spin = 0;
  } else if (emitter.rotType === 'RANDOM') {
    slot.rot = THREE.MathUtils.degToRad(runtime.rng.next() * 360);
    slot.spin = 0;
  } else {
    slot.rot = THREE.MathUtils.degToRad(emitter.initRot + runtime.rng.centered(emitter.initRotRnd));
    slot.spin = THREE.MathUtils.degToRad(emitter.spin + runtime.rng.centered(Math.abs(emitter.spinRnd)));
  }

  slot.baseSize = Math.max(0.001, emitter.sizeX + runtime.rng.centered(emitter.sizeRnd));
  slot.col = { ...sampleColor(emitter.colorKeys, 0) };
  slot.frame = emitter.uvAnim === 'rand' ? Math.floor(runtime.rng.next() * emitter.sheetCols * emitter.sheetRows) : 0;
}

export function updateRuntime(runtime: RuntimeEmitter, dt: number, playing: boolean, lodFactor = 1): number {
  const emitter = runtime.emitter;
  if (playing) {
    if (emitter.loop && emitter.cycle > 0 && runtime.localTime >= emitter.cycle) {
      runtime.localTime = 0;
      runtime.spawnAcc = 0;
    }
    runtime.localTime += dt;

    const delay = emitter.delay || 0;
    if (runtime.localTime >= delay) {
      if (emitter.burst > 0 && runtime.localTime - dt < delay + dt && runtime.localTime >= delay) {
        for (let i = 0; i < emitter.burst; i += 1) spawnParticle(runtime);
      }

      if (emitter.loop || runtime.localTime <= emitter.cycle) {
        const effectiveRate = emitter.rate * lodFactor;
        runtime.spawnAcc += effectiveRate * dt;
        const count = Math.floor(runtime.spawnAcc);
        runtime.spawnAcc -= count;
        for (let i = 0; i < Math.min(count, 16); i += 1) spawnParticle(runtime);
      }
    }
  }

  const totalFrames = Math.max(1, emitter.sheetCols * emitter.sheetRows);
  const drag = Math.max(0, 1 - emitter.drag * dt);
  let alive = 0;

  for (let i = 0; i < runtime.particles.length; i++) {
    const particle = runtime.particles[i];
    if (!particle.alive) continue;
    particle.age += dt;
    if (particle.age >= particle.life) {
      particle.alive = false;
      runtime.freeSlots.push(i);
      runtime.aliveCount--;
      continue;
    }

    alive += 1;
    const t = particle.age / particle.life;
    const speedMul = sampleCurve(emitter.speedCurve, t);
    particle.vy += emitter.gravity * dt;
    particle.vx += emitter.windX * dt;
    particle.vz += emitter.windZ * dt;
    particle.vx *= drag;
    particle.vy *= drag;
    particle.vz *= drag;

    if (emitter.turb && emitter.turb !== 0) {
      const phase = runtime.localTime * (emitter.turbFreq || 1);
      const tx = Math.sin(phase * 1.7 + particle.px * 3.1 + i * 0.3) * emitter.turb;
      const ty = Math.sin(phase * 2.1 + particle.py * 2.7 + i * 0.7) * emitter.turb * 0.5;
      const tz = Math.cos(phase * 1.9 + particle.pz * 3.3 + i * 0.5) * emitter.turb;
      particle.vx += tx * dt;
      particle.vy += ty * dt;
      particle.vz += tz * dt;
    }

    if (emitter.attractorStr && emitter.attractorStr !== 0) {
      const ay = emitter.attractorY ?? 0;
      const bx = 0, bz = 0;
      const dax = bx - particle.px;
      const day = ay - particle.py;
      const daz = bz - particle.pz;
      const dLen = Math.sqrt(dax * dax + day * day + daz * daz) + 0.001;
      const aF = emitter.attractorStr * dt / dLen;
      particle.vx += dax * aF;
      particle.vy += day * aF;
      particle.vz += daz * aF;
    }

    particle.px += particle.vx * speedMul * dt;
    particle.py += particle.vy * speedMul * dt;
    particle.pz += particle.vz * speedMul * dt;

    if (emitter.groundBounce && particle.py < 0 && particle.vy < 0) {
      particle.py = 0;
      particle.vy = -particle.vy * emitter.bounceFac;
      particle.vx *= 0.85;
      particle.vz *= 0.85;
    }

    particle.rot += particle.spin * sampleCurve(emitter.spinCurve, t) * dt;
    particle.col = sampleColor(emitter.colorKeys, t);

    if (emitter.velStretch && emitter.velStretch !== 0) {
      const sp = Math.sqrt(particle.vx * particle.vx + particle.vy * particle.vy + particle.vz * particle.vz);
      if (sp > 0.01) {
        particle.stretchRot = Math.atan2(particle.vx, particle.vz);
        particle.stretch = emitter.velStretch;
      } else {
        particle.stretch = 0;
      }
    }

    if (emitter.uvAnim === 'loop') particle.frame = Math.floor(particle.age * emitter.animFPS) % totalFrames;
    else if (emitter.uvAnim === 'once') particle.frame = Math.min(Math.floor(particle.age * emitter.animFPS), totalFrames - 1);
    else if (emitter.uvAnim === 'life') particle.frame = clamp(Math.floor(t * totalFrames), 0, totalFrames - 1);
  }

  runtime.dirty = false;
  return alive;
}

File: Metin2AssetStudioCpp/src/runtime/ParticleSimulation.cpp
#include "ParticleSimulation.h"
#include "utils/MathUtils.h"
#include "utils/CurveUtils.h"
#include <cmath>
#include <algorithm>
#include <cstdint>

static const double PI = 3.14159265358979323846;

std::vector<Particle> ParticleSimulation::createParticles(int count) {
    std::vector<Particle> p(count);
    for (int i = 0; i < count; ++i) {
        p[i].alive = false;
        p[i].age = 0;
        p[i].life = 1;
        p[i].px = 0; p[i].py = 0; p[i].pz = 0;
        p[i].vx = 0; p[i].vy = 0; p[i].vz = 0;
        p[i].rot = 0;
        p[i].spin = 0;
        p[i].baseSize = 1;
        p[i].col.r = 1; p[i].col.g = 1; p[i].col.b = 1; p[i].col.a = 1;
        p[i].frame = 0;
        p[i].boneOx = 0; p[i].boneOy = 0; p[i].boneOz = 0;
        p[i].stretchRot = 0;
        p[i].stretch = 0;
    }
    return p;
}

void ParticleSimulation::resetRuntime(RuntimeEmitter& runtime) {
    for (auto& p : runtime.particles) {
        p.alive = false;
    }
    runtime.freeSlots.clear();
    for (int i = (int)runtime.particles.size() - 1; i >= 0; --i) {
        runtime.freeSlots.push_back(i);
    }
    runtime.spawnAcc = 0;
    runtime.localTime = 0;
    runtime.rng = SeededRandom(runtime.emitter->uid * 2654435761ULL);
    runtime.dirty = false;
}

RuntimeEmitter ParticleSimulation::createRuntime(Emitter* emitter) {
    RuntimeEmitter rt;
    rt.emitter = emitter;
    rt.particles = createParticles(emitter->maxP);
    resetRuntime(rt);
    return rt;
}

void ParticleSimulation::spawnParticle(RuntimeEmitter& runtime) {
    if (runtime.freeSlots.empty()) return;
    int idx = runtime.freeSlots.back();
    runtime.freeSlots.pop_back();
    Particle& p = runtime.particles[idx];
    p.alive = true;
    p.age = 0;
    Emitter& e = *runtime.emitter;
    SeededRandom& rng = runtime.rng;

    p.life = std::max(0.01, e.life + rng.centered(e.lifeRnd));

    double radius = e.shapeRadius;
    double r = radius;
    switch (e.shape) {
    case ShapeType::Point:
        p.px = 0; p.py = 0; p.pz = 0;
        break;
    case ShapeType::Box:
        p.px = rng.centered(r * 0.7);
        p.py = rng.centered(r * 0.7);
        p.pz = rng.centered(r * 0.7);
        break;
    case ShapeType::Sphere:
    case ShapeType::SphereVol: {
        double theta = rng.next() * 2.0 * PI;
        double phi = std::acos(2.0 * rng.next() - 1.0);
        double dist = (e.shape == ShapeType::SphereVol) ? std::cbrt(rng.next()) * r : r;
        p.px = dist * std::sin(phi) * std::cos(theta);
        p.py = dist * std::cos(phi);
        p.pz = dist * std::sin(phi) * std::sin(theta);
        break;
    }
    case ShapeType::Ring:
    case ShapeType::Disc:
    case ShapeType::Cone: {
        double angle = rng.next() * 2.0 * PI;
        double dist = (e.shape == ShapeType::Point || e.shape == ShapeType::Ring) ? r : std::sqrt(rng.next()) * r;
        if (e.shape == ShapeType::Ring) dist = r;
        double height = (e.shape == ShapeType::Cone) ? rng.next() * r : 0;
        p.px = dist * std::cos(angle);
        p.py = height;
        p.pz = dist * std::sin(angle);
        break;
    }
    }

    double dirYawRad = MathUtils::degToRad(e.dirYaw + 90.0);
    double dirPitchRad = MathUtils::degToRad(e.dirPitch);
    double cx = std::sin(dirYawRad) * std::cos(-dirPitchRad);
    double cy = std::sin(-dirPitchRad);
    double cz = std::cos(dirYawRad) * std::cos(-dirPitchRad);
    MathUtils::Vec3 dir(cx, cy, cz);
    dir.normalize();

    double spreadRad = MathUtils::degToRad(e.spread);
    if (spreadRad > 0.001) {
        double theta = rng.next() * 2.0 * PI;
        double phi = rng.next() * spreadRad;
        double sx = std::sin(theta) * std::sin(phi);
        double sy = std::cos(theta) * std::sin(phi);
        double sz = std::cos(phi);
        MathUtils::Vec3 axis(rng.centered(1), rng.centered(1), rng.centered(1));
        axis.normalize();
        double c = sz;
        double s = std::sin(phi);
        double oneC = 1.0 - c;
        double m00 = oneC * axis.x * axis.x + c;
        double m01 = oneC * axis.x * axis.y - axis.z * s;
        double m02 = oneC * axis.x * axis.z + axis.y * s;
        double m10 = oneC * axis.x * axis.y + axis.z * s;
        double m11 = oneC * axis.y * axis.y + c;
        double m12 = oneC * axis.y * axis.z - axis.x * s;
        double m20 = oneC * axis.x * axis.z - axis.y * s;
        double m21 = oneC * axis.y * axis.z + axis.x * s;
        double m22 = oneC * axis.z * axis.z + c;
        double dx = m00 * dir.x + m01 * dir.y + m02 * dir.z;
        double dy = m10 * dir.x + m11 * dir.y + m12 * dir.z;
        double dz = m20 * dir.x + m21 * dir.y + m22 * dir.z;
        dir = MathUtils::Vec3(dx, dy, dz);
        dir.normalize();
    }

    double spd = e.speed + rng.centered(e.speedRnd);
    p.vx = dir.x * spd;
    p.vy = dir.y * spd;
    p.vz = dir.z * spd;

    switch (e.rotType) {
    case RotType::None:
        p.rot = 0;
        break;
    case RotType::Random:
        p.rot = rng.next() * 2.0 * PI;
        break;
    case RotType::Spin:
        p.rot = MathUtils::degToRad(e.initRot + rng.centered(e.initRotRnd));
        break;
    }
    p.spin = MathUtils::degToRad(e.spin);

    p.baseSize = e.sizeX + rng.centered(e.sizeRnd);

    p.col.r = 1; p.col.g = 1; p.col.b = 1; p.col.a = 1;
    if (!e.colorKeys.empty()) {
        ColorKey ck = CurveUtils::sampleColor(e.colorKeys, 0);
        p.col.r = ck.r;
        p.col.g = ck.g;
        p.col.b = ck.b;
        p.col.a = ck.a;
    }

    if (e.uvAnim == UVAnimType::Rand) {
        p.frame = (int)(rng.next() * (e.sheetCols * e.sheetRows));
    } else {
        p.frame = 0;
    }

    ++runtime.aliveCount;
}

int ParticleSimulation::updateRuntime(RuntimeEmitter& runtime, double dt, bool playing, double lodFactor) {
    SeededRandom& rng = runtime.rng;
    Emitter& e = *runtime.emitter;
    double effectiveDt = dt;
    bool firstFrame = false;

    if (!playing) {
        return runtime.aliveCount;
    }

    if (runtime.localTime == 0) {
        firstFrame = true;
    }

    if (e.delay > 0 && runtime.localTime < e.delay) {
        double newTime = runtime.localTime + effectiveDt;
        if (newTime >= e.delay) {
            effectiveDt = newTime - e.delay;
        } else {
            runtime.localTime = newTime;
            return runtime.aliveCount;
        }
    }

    runtime.localTime += effectiveDt;

    double cycleDur = e.cycle;
    if (e.loop && cycleDur > 0) {
        if (runtime.localTime >= cycleDur) {
            runtime.localTime = std::fmod(runtime.localTime, cycleDur);
            for (auto& p : runtime.particles) {
                if (p.alive) {
                    p.alive = false;
                    runtime.freeSlots.push_back((int)(&p - &runtime.particles[0]));
                    --runtime.aliveCount;
                }
            }
            runtime.spawnAcc = 0;
        }
    }

    if (firstFrame || (e.burst > 0 && runtime.localTime < dt * 2)) {
        int burstCount = e.burst;
        for (int i = 0; i < burstCount; ++i) {
            spawnParticle(runtime);
        }
    }

    if (e.rate > 0) {
        runtime.spawnAcc += effectiveDt * e.rate * lodFactor;
        while (runtime.spawnAcc >= 1.0) {
            spawnParticle(runtime);
            runtime.spawnAcc -= 1.0;
        }
    }

    double dtFactor = 1.0;
    for (auto& p : runtime.particles) {
        if (!p.alive) continue;
        p.age += effectiveDt * dtFactor;

        if (p.age >= p.life) {
            p.alive = false;
            runtime.freeSlots.push_back((int)(&p - &runtime.particles[0]));
            --runtime.aliveCount;
            continue;
        }

        double t = std::min(1.0, p.age / p.life);

        double speedFac = CurveUtils::sampleCurve(e.speedCurve, t);
        double spdX = p.vx * speedFac;
        double spdY = p.vy * speedFac;
        double spdZ = p.vz * speedFac;

        spdY += e.gravity * effectiveDt;

        spdX += e.windX * effectiveDt;
        spdZ += e.windZ * effectiveDt;

        if (e.drag > 0) {
            double dragFac = 1.0 / (1.0 + e.drag * effectiveDt);
            spdX *= dragFac;
            spdY *= dragFac;
            spdZ *= dragFac;
        }

        if (e.turb > 0) {
            double freq = e.turbFreq > 0 ? e.turbFreq : 1.0;
            double phaseX = p.px * freq + p.age * 2.0;
            double phaseY = p.py * freq + p.age * 1.7;
            double phaseZ = p.pz * freq + p.age * 2.3;
            double turbStr = e.turb * effectiveDt;
            spdX += std::sin(phaseX) * turbStr;
            spdY += std::sin(phaseY) * turbStr;
            spdZ += std::cos(phaseZ) * turbStr;
        }

        if (e.attractorStr != 0) {
            double ax = -p.px * e.attractorStr * effectiveDt;
            double ay = (e.attractorY - p.py) * e.attractorStr * effectiveDt;
            double az = -p.pz * e.attractorStr * effectiveDt;
            spdX += ax;
            spdY += ay;
            spdZ += az;
        }

        p.vx = spdX;
        p.vy = spdY;
        p.vz = spdZ;

        p.px += p.vx * effectiveDt;
        p.py += p.vy * effectiveDt;
        p.pz += p.vz * effectiveDt;

        if (e.groundBounce && p.py <= 0) {
            p.py = 0;
            p.vy = -p.vy * e.bounceFac;
        }

        if (e.rotType == RotType::Spin) {
            double spinFac = CurveUtils::sampleCurve(e.spinCurve, t);
            p.rot += p.spin * spinFac * effectiveDt;
        }

        ColorKey ck = CurveUtils::sampleColor(e.colorKeys, t);
        p.col.r = ck.r;
        p.col.g = ck.g;
        p.col.b = ck.b;
        p.col.a = ck.a;

        double alphaFac = CurveUtils::sampleCurve(e.alphaCurve, t);
        p.col.a *= alphaFac;

        double szFac = CurveUtils::sampleCurve(e.sizeCurve, t);
        double sz = p.baseSize * szFac;
        if (e.sizeNonUniform) {
            double szY = (e.sizeY + rng.centered(e.sizeRnd)) * szFac;
            p.boneOy = szY;
        } else {
            p.boneOy = sz;
        }
        p.boneOx = sz;
        p.boneOz = sz;

        if (e.velStretch > 0) {
            double velMag = std::sqrt(p.vx * p.vx + p.vy * p.vy + p.vz * p.vz);
            double stretch = velMag * e.velStretch * effectiveDt;
            p.stretch = stretch;
            if (velMag > 0.001) {
                p.stretchRot = std::atan2(p.vz, p.vx);
            }
        }

        if (e.uvAnim != UVAnimType::Rand && e.uvAnim != UVAnimType::None) {
            int totalFrames = e.sheetCols * e.sheetRows;
            if (totalFrames > 1) {
                switch (e.uvAnim) {
                case UVAnimType::Loop:
                    p.frame = (int)((p.age * e.animFPS) / 1.0) % totalFrames;
                    break;
                case UVAnimType::Once: {
                    int f = (int)((p.age / p.life) * totalFrames);
                    p.frame = std::min(f, totalFrames - 1);
                    break;
                }
                case UVAnimType::Life:
                    p.frame = (int)((p.age / p.life) * totalFrames) % totalFrames;
                    break;
                default:
                    break;
                }
            }
        }
    }

    return runtime.aliveCount;
}
